assets/js/plugins.coffee

```diff
@@ -0,0 +1,99 @@
+if module?
+  global._ = require('lodash')
+  global.Logger = require('./logger.coffee')
+  global.DependencyGraph = require('dependencies-online')
+
+(() ->
+  class PluginDatabaseAccess
+    constructor: (@data, @pluginName) ->
+
+    getVersion: () ->
+      @data.store.getPluginVersion @pluginName
+    setVersion: (version) ->
+      @data.store.setPluginVersion @pluginName, version
+    setRowData: (row, args...) ->
+      if args.length == 1
+        [key, [data]] = ['data', args]
+      else if args.length == 2
+        [key, data] = args
+      @data.store.setPluginRowData @pluginName, row.id, key, data
+    getRowData: (row, key='data') ->
+      @data.store.getPluginRowData @pluginName, row.id, key
+    setGlobalData: (key, args...) ->
+      if args.length == 1
+        [key, [data]] = ['data', args]
+      else if args.length == 2
+        [key, data] = args
+      @data.store.setPluginGlobalData @pluginName, key, data
+    getGlobalData: (key='data') ->
+      @data.store.getPluginGlobalData @pluginName, key
+
+    transformRowData: (row, args...) ->
+      if args.length == 1
+        [key, [transform]] = ['data', args]
+      else if args.length == 2
+        [key, transform] = args
+      @setRowData row, key, (transform (@getRowData row, key))
+    transformGlobalData: (args...) ->
+      if args.length == 1
+        [key, [transform]] = ['data', args]
+      else if args.length == 2
+        [key, transform] = args
+      @setGlobalData key, (transform (@getGlobalData key))
+  
+  class PluginAPI
+    constructor: (@view) ->
+      @plugins = {}
+      @pluginClass = {}
+      @metadata = {}
+      @pluginDependencies = new DependencyGraph()
+      # Convenience accessors
+      @cursor = @view.cursor
+
+    registerPlugin: (plugin) ->
+      # Validate metadata before accepting registration
+      metadata = plugin.metadata
+      unless metadata?
+        throw "Plugin is missing all metadata"
+      required_props = ['name', 'version', 'stores_data']
+      if (_.result metadata, 'stores_data')
+        required_props.push 'data_version'
```

Comment:
might also be cleaner to have stores_data be always treated as true, and for data_version to basically default to 1 (or 0)